modules/fedimint-ecash-migration-server/src/lib.rs

```diff
@@ -227,20 +435,321 @@ impl ServerModule for EcashMigration {
+
+        let spend_book_hash = hash_spend_book(
+            dbtx.find_by_prefix(&OriginSpendBookTransferPrefix { transfer_id })
+                .await
+                .map(|(OriginSpendBookKey { nonce, .. }, ())| nonce),
+        )
+        .await;
+
+        if spend_book_hash != transfer_metadata.origin_spend_book_hash {
+            return Err(ApiError::bad_request(format!(
+                "Spend book hash does not match pre-committed hash (expected: {}; got: {})",
+                transfer_metadata.origin_spend_book_hash, spend_book_hash,
+            )));
+        }
+
+        Ok(spend_book_hash)
+    }
+
+    async fn get_transfer_status(
+        &self,
+        dbtx: &mut DatabaseTransaction<'_>,
+        transfer_id: TransferId,
+    ) -> Result<GetTransferStatusResponse, ApiError> {
+        let transfer_metadata = dbtx
+            .get_value(&TransferMetadataKey(transfer_id))
+            .await
+            .ok_or(ApiError::not_found("Transfer not found".to_owned()))?;
+
+        let is_active = is_transfer_active(dbtx, transfer_id, self.num_peers).await;
+
+        let deposited_amount = dbtx
+            .get_value(&DepositedAmountKey(transfer_id))
+            .await
+            .unwrap_or(Amount::ZERO);
+
+        let withdrawn_amount = dbtx
+            .get_value(&WithdrawnAmountKey(transfer_id))
+            .await
+            .unwrap_or(Amount::ZERO);
+
+        Ok(GetTransferStatusResponse {
+            is_active,
+            spend_book_merkle_root: transfer_metadata.origin_spend_book_merkle_root,
+            key_set_hash: transfer_metadata.origin_key_set_hash,
+            total_entries: transfer_metadata.num_spend_book_entries,
+            total_amount: deposited_amount,
+            redeemed_amount: withdrawn_amount,
+        })
     }
 }
+
+/// Check if a transfer is active (has enough votes)
+async fn is_transfer_active(
+    dbtx: &mut DatabaseTransaction<'_>,
+    transfer_id: TransferId,
+    num_peers: NumPeers,
+) -> bool {
+    let vote_count = dbtx
+        .find_by_prefix(&ActivationVoteTransferPrefix { transfer_id })
+        .await
+        .count()
+        .await;
+
+    vote_count == num_peers.total()
+}
+
+async fn get_next_transfer_id(dbtx: &mut DatabaseTransaction<'_>) -> TransferId {
+    let max_id = dbtx
+        .find_by_prefix_sorted_descending(&TransferMetadataKeyPrefix)
+        .await
+        .next()
+        .await
+        .map_or(0, |(TransferMetadataKey(id), _)| id.0);
+
+    TransferId(max_id + 1)
+}
+
+async fn check_auth(
+    context: &mut ApiEndpointContext<'_>,
+    transfer_id: TransferId,
+    request: &impl Encodable,
+) -> Result<(), ApiError> {
+    let transfer_metadata = context
+        .dbtx()
+        .get_value(&TransferMetadataKey(transfer_id))
+        .await
+        .ok_or(ApiError::not_found("Transfer not found".to_owned()))?;
+    let creator_keys = transfer_metadata.creator_keys;
+
+    let auth = context.request_auth().ok_or(ApiError::unauthorized())?;
```

Comment:
Oh... so we ... use the request_auth ... Yyyy... Feels like mixing different concerns etc. and asking for trouble down the line. Would rather take the signature as a separate argument. Could be `struct SignedRequest<RequestT> { inner: RequestT, signature: NaiveThresholdSignature }` for convenience.